cmd/containerboot,kube,util/linuxfw: configure kube egress proxies to route to 1+ tailnet targets

irbekrm opened the pull request:
This is first part of the work to allow running multiple replicas of the Kubernetes operator _egress_ proxies per tailnet service + to allow exposing multiple tailnet services via each proxy replica.

This expands the existing iptables/nftables-based proxy configuration mechanism.

A proxy can now be configured to route to one or more tailnet targets via a (mounted) config file that, for each tailnet target, specifies:
- the target's tailnet IP or FQDN
- mappings of container ports to which cluster workloads will send traffic to tailnet target ports where the traffic should be forwarded.

Example configfile contents (this will be created by the operator, not users):
```json
{
  "some-svc": {"tailnetTarget":{"fqdn":"foo.tailnetxyz.ts.net","ports"{"tcp:4006:80":{"protocol":"tcp","matchPort":4006,"targetPort":80},"tcp:4007:443":{"protocol":"tcp","matchPort":4007,"targetPort":443}}}}
}
```
A proxy that is configured with this config file will configure firewall rules to route cluster traffic to the tailnet targets. It will then watch the config file for updates as well as monitor relevant netmap updates and reconfigure firewall as needed.

This adds a bunch of new iptables/nftables functionality to make it easier to dynamically update the firewall rules without needing to restart the proxy Pod as well as to make it easier to debug/understand the rules:

- for iptables, each portmapping is a DNAT rule with a comment pointing at the 'service',i.e:

```
-A PREROUTING ! -i tailscale0 -p tcp -m tcp --dport 4006 -m comment --comment "some-svc:tcp:4006 -> tcp:80" -j DNAT --to-destination 100.64.1.18:80
```
Additionally there is a SNAT rule for each tailnet target, to mask the source address.

- for nftables, a separate prerouting chain is created for each tailnet target and all the portmapping rules are placed in that chain. This makes it easier to look up rules and delete services when no longer needed. (nftables allows hooking a custom chain to a prerouting hook, so no extra work is needed to ensure that the rules in the service chains are evaluated).

The next steps will be to get the Kubernetes Operator to generate the configfile and ensure it is mounted to the relevant proxy nodes.

Updates tailscale/tailscale#13406

irbekrm commented on util/linuxfw/nftables_for_svcs.go:
```diff
@@ -0,0 +1,245 @@
+		return fmt.Errorf("error ensuring chain for %s: %w", svc, err)
+	}
+	meta := svcPortMapRuleMeta(svc, targetIP, pm)
+	rule, err := n.findRuleByMetadata(t, ch, meta)
```
There is another util function to search for a rule https://github.com/tailscale/tailscale/blob/3e9ca6c64bcd41c37ca73a370e2928b367e04781/util/linuxfw/nftables_runner.go#L707
The one I am adding should be more efficient and reliable for our use case as it searches by user data field value.
I would also like to change our existing proxy rules to create-if-not-exists and use this to search if a rule already exists.

irbekrm commented on kube/egressservices/egressservices.go:
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package egressservices contains shared types for exposing tailnet services to
```
These configs will only ever be created by the Kubernetes operator, not users directly

irbekrm commented on util/linuxfw/nftables_runner.go:
```diff
@@ -569,6 +569,12 @@ type NetfilterRunner interface {
 	// the Tailscale interface, as used in the Kubernetes egress proxies.
 	DNATNonTailscaleTraffic(exemptInterface string, dst netip.Addr) error
 
+	EnsurePortMapRuleForSvc(svc, tun string, targetIP netip.Addr, pm PortMap) error
+
+	DeletePortMapRuleForSvc(svc, tun string, targetIP netip.Addr, pm PortMap) error
+
+	DeleteSvc(svc, tun string, targetIPs []netip.Addr, pm []PortMap) error
+
```
cc @raggi  @KevinLiang10 - just an fyi that this adds a couple more methods to the netfilter runner, these are only used for the egress proxies, no changes for the router functionality
At the moment I think there is a clean split between the iptables/nftables methods used for the router and the kube proxies, but also for future we are considering implementing the kube TCP proxy in code, so I am not giving much thought to whether we want to put this functionality somewhere else.

tomhjp reviewed (commented):
This is amazing! I left a bunch of comments but they are mostly nits/naming/structure. The actual implementation and test coverage looks :chefs-kiss:. Especially the nftables implementation which looks like it was extra challenging

irbekrm commented on cmd/containerboot/main.go:
```diff
@@ -593,6 +598,30 @@ runLoop:
+						ep := egressProxy{
+							cfgPath:      cfg.EgressServicesPath,
+							nfr:          nfr,
+							kc:           kc,
```
Definitely TBD as part of refactor of containerboot 👍🏼

irbekrm commented on cmd/containerboot/services.go:
```diff
@@ -0,0 +1,568 @@
+		return false
+	}
+
+	// If proxy's tailnet addresses have changed, resync.
```
It also needs to mask the source IP from workload Pod's IP to its own tailnet IP address to receive response packets, so we need to resync if its addresses have changed

irbekrm commented on cmd/containerboot/services.go:
```diff
@@ -0,0 +1,568 @@
+	var eventChan <-chan fsnotify.Event
+	// TODO (irbekrm): take a look if this can be pulled into a single func
+	// shared with serve config loader.
+	if w, err := fsnotify.NewWatcher(); err != nil {
```
> Under what conditions do we expect this to fail

I pulled this behaviour from the serve watch and right now I don't remember if we had someone report that the watch was failing for certain kube setups or whether we assumed it might be so.
The assumption in any case is that watcher is unable to start for certain type of containers at all, so we fall back to the ticker. We can investigate and add a comment when/if that would happen, but would rather keep this behaviour till then

irbekrm commented on cmd/containerboot/services.go:
```diff
@@ -0,0 +1,568 @@
+// deleteUnneccessaryServices ensure that any services found on status, but not
+// present in config are deleted.
+func (ep *egressProxy) deleteUnnecessaryServices(cfgs *egressservices.Configs, status *egressservices.Status) error {
+	if !hasServicesConfigured(status) {
```
As discussed earlier-  status being keyed by Pod IP should be sufficient as we anyway crash the Pod on firewall configuration errors(=IP change). I'll add a comment for that though as we might want to reconsider the failure behaviour in the future

irbekrm commented on kube/egressservices/egressservices.go:
```diff
@@ -0,0 +1,72 @@
+	// Ports contains mappings for ports that can be accessed on the tailnet
+	// target keyed by a predictable name for easier lookup.
+	// {"tcp:80:4003":{"protocol":"tcp","src":80,"dst":4003}}
+	Ports map[PortMapName]PortMap `json:"ports"`
```
> Both PortMapName and PortMap look like they have the same information in them

They do, but I don't want to loop over and compare array items to find a port- especially in the follow up PRs where I need to compare a port as defined here with a port as it is on a Service spec with additional fields potentially added by kube- so lookup by name that includes things we care about makes it easier

tomhjp replied:
Sorry I probably didn't explain this well - you can use the `PortMap` type as the map key instead of having to create a string-type purely for the purpose of being the map key, because the `PortMap` type is a comparable struct. I pushed 005b2e2f22db4a68df9e0bb78bf74e0de0e6e93d to demonstrate how this _slightly_ simplifies the code. Very much a nit, I mostly commented because I wondered if I was missing something.

tomhjp replied:
And I don't entirely know where this is going in the next PR, so I may be missing the reason why that struct will no longer work as a key in the near future.

irbekrm replied:
@tomhjp this would be a lot nicer, but we also need to encode this as JSON and afaiu JSON does not support objects as map key types. I think it's probably possible via a custom encoder or something, but in that case I'm not sure if it's simpler any more

irbekrm replied:
Seems like I could convert the map if it had struct keys to json [via a custom `MarshalText` method for the struct](https://pkg.go.dev/encoding/json#Marshal:~:text=Map%20values%20encode%20as%20JSON%20objects.%20The%20map%27s%20key%20type%20must%20either%20be%20a%20string%2C%20an%20integer%20type%2C%20or%20implement%20encoding.TextMarshaler.). That would still be fine
But then I would also need to parse those strings back into structs when converting json back to go and that seems a bit messy

irbekrm replied:
I'll try it out though

irbekrm replied:
Yeah, with the custom marshalling this still seems like an improvement, thanks

https://github.com/tailscale/tailscale/pull/13531/commits/70fc37363d7f02eba72dc771fda74e4d31cd0804#diff-62d368cb3988ff2fc5340b71425899638d96b4df7d0551de88f3f6db4e174a17R55-R84

tomhjp replied:
Ah I missed the JSON serialisation requirement. It doesn't feel right to me that the requirements of the code leak into the requirements of the serialisation and vice-versa.

I think the serialised form is bested suited to being a simple list of objects, as that most closely represents how that config gets turned into real nftable/iptable rules, and is very extensible for any future updates we make (including adding a version field in those objects). We can separately handle converting that list of objects into a set as we parse the config file - no need to tie it up in the JSON marshalling interface.

irbekrm commented on kube/kubeclient/client.go:
```diff
@@ -257,7 +257,7 @@ type JSONPatch struct {
 	for _, p := range patch {
-		if p.Op != "remove" && p.Op != "add" {
+		if p.Op != "remove" && p.Op != "add" && p.Op != "replace" {
 			panic(fmt.Errorf("unsupported JSON patch operation: %q", p.Op))
```
👍🏼 will add a TODO to potentially remove. I didn't add this bit, so it might be either that it was added in error or there is some reason I can't think of now

irbekrm replied:
Actually I really can't see why we would need to panic, so returning an error now

irbekrm replied:
Thanks for spotting this

irbekrm commented on util/linuxfw/iptables_for_svcs.go:
```diff
@@ -0,0 +1,79 @@
+// EnsurePortMapRuleForSvc adds a prerouting rule that forwards traffic received
+// on match port and NOT on the provided interface to target IP and target port.
+// Rule will only be added if it does not already exists.
+func (i *iptablesRunner) EnsurePortMapRuleForSvc(svc, tun string, targetIP netip.Addr, pm PortMap) error {
```
fwiw the logic is the same as we have for non-HA egress today https://github.com/tailscale/tailscale/blob/77832553e538618ad94c6fc33f32204c3dcb552d/util/linuxfw/iptables_runner.go#L381

irbekrm commented on util/linuxfw/nftables_runner.go:
```diff
@@ -569,6 +569,12 @@ type NetfilterRunner interface {
 	// the Tailscale interface, as used in the Kubernetes egress proxies.
 	DNATNonTailscaleTraffic(exemptInterface string, dst netip.Addr) error
 
+	EnsurePortMapRuleForSvc(svc, tun string, targetIP netip.Addr, pm PortMap) error
```
This is the same 'service' concept as used across the whole of this PR- the idea is that there is an 'egress service' and rules are set up for such a service. The first arg here is the name of this 'service'- I don't think we should rename that to 'description'

irbekrm commented on util/linuxfw/nftables_runner.go:
```diff
@@ -569,6 +569,12 @@ type NetfilterRunner interface {
 	// the Tailscale interface, as used in the Kubernetes egress proxies.
 	DNATNonTailscaleTraffic(exemptInterface string, dst netip.Addr) error
 
+	EnsurePortMapRuleForSvc(svc, tun string, targetIP netip.Addr, pm PortMap) error
```
I have used 'targetIP' across all methods for services because it matches 'targetPort' else we'd have 'dst' for the tailnet target IP and 'targetPort' for tailnet target port

tomhjp replied:
That's a valid argument for the call stack to get to these methods. I think there is also an argument to be more consistent with the other methods in this interface, but I don't feel that strongly either way 👍

irbekrm replied:
> I think there is also an argument to be more consistent with the other methods in this interface

I was actually considering changing 'dst' to 'target' for that reason across the methods, but didn't do in this PR (the amount of changes is already significant). Also, we don't know yet what we want to do with this configuration mechanism in future (implement in userspace, bring into the router, etc)

irbekrm commented on cmd/containerboot/services.go:
```diff
@@ -0,0 +1,568 @@
+
+//go:build linux
+
+package main
```
I would prefer to put this into a package once we start doing ingress HA if you don't mind

I imagine that we will have some sort of ./proxy library with proxy service related bits for both types and later we can add all the other proxying logic there too.
I find that it's easier to split something out once there is more than one implementation of it as then it's clearer which are the shared bits - and it's easier to design a package hierarchy for related things like ingress/egress HA services out of flat files than if it was already partially across different packages

irbekrm commented on cmd/containerboot/services.go:
```diff
@@ -0,0 +1,568 @@
+	protocol      string
+}
+
+func wantsServicesConfigured(cfgs *egressservices.Configs) bool {
```
Yeah that makes sense. I was wanting to re-use some of `egressservices` functionality in netfilter runner so tried to keep it light, but didn't do that anyway so makes sense to move

irbekrm replied:
hmm, but thinking about it, half of those are a nil check that I anyway cannot do there

irbekrm replied:
ok, I don't want to make them methods to be able to keep the nil checks in and I don't want to just port them over as functions unless later down the line we need the same functionality in other packages. If you don't mind I'll leave it as is

tomhjp replied:
Technically it's perfectly [valid Go](https://go.dev/play/p/CnOKEa20WYP) to do a nil-check in a pointer receiver method. Having said that, I don't think it's a good pattern so I am fine with that conclusion 👍

irbekrm commented on cmd/containerboot/services.go:
```diff
@@ -0,0 +1,568 @@
+	}
+
+	if err := ep.sync(ctx, n); err != nil {
+		log.Fatal(err)
```
We should refactor this in future.

For now- I've moved the log.Fatal into the main function

irbekrm commented:
@tomhjp thank you very much for the review.

I've fixed the smaller bits, but not the refactor/package split. I added a comment for this, but TL;DR I'd like to do it once we start looking at also adding ingress HA- so that we can look at how to structure libraries for a different type of 'proxy'. Let me know if you agree

tomhjp reviewed (approved):
LGTM! Thanks for all the explanations on the bits I didn't understand. I am fine with deferring any package-related refactoring 👍
